### PR TITLE
tools/cephfs/first-damage: repair specific first=CEPH_NOSNAP damage

### PR DESCRIPTION
**Pull request by batrick**
## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/tools/cephfs/first-damage.py [hunk 1]
```diff
@@ -76,11 +79,19 @@ def traverse(MEMO, ioctx):
...
+                log.debug('\t%s: val size %d', dnk, len(val))
                 (first,) = struct.unpack('<I', val[:4])
                 if first > NEXT_SNAP:
                     log.warning(f"found {o.key}:{dnk} first (0x{first:x}) > NEXT_SNAP (0x{NEXT_SNAP:x})")
-                    if REMOVE:
+                    if REPAIR_NOSNAP and dnk.endswith("_head") and first == CEPH_NOSNAP:
+                        log.warning(f"repairing first==CEPH_NOSNAP damage, setting to NEXT_SNAP (0x{NEXT_SNAP:x})")
+                        first = NEXT_SNAP
+                        nval = bytearray(val)
+                        struct.pack_into("<I", nval, 0, NEXT_SNAP)
```
Comment (vshankar) [inline src/tools/cephfs/first-damage.py:90]: This looks right.

@batrick Help me remind, we never tested in teuthology, isn't it? (although the test script is under qa/workunits).

**Comment by gregsfortytwo** [comment]
LGTM.

But do we have any idea how this can happen? This seems to be a different kind of problem than the other one triggered by mysql.

**Comment by vshankar** [comment]
> LGTM.
> 
> But do we have any idea how this can happen? This seems to be a different kind of problem than the other one triggered by mysql.

This instance did not have postgresql or any kind of database - the corruption happened to a random pdf or xml file :/

**Comment by gregsfortytwo** [comment]
> > LGTM.
> > But do we have any idea how this can happen? This seems to be a different kind of problem than the other one triggered by mysql.
> 
> This instance did not have postgresql or any kind of database - the corruption happened to a random pdf or xml file :/

Yeah, but it was like 10% of files in the system, right? That seems like a bad logic path applying incorrect updates, not a memory corruption...

File: src/tools/cephfs/first-damage.py (hunk 1 above)
Comment (batrick) [reply src/tools/cephfs/first-damage.py:90]: No, we don't. It's just a script for QE to run.

**Review by vshankar — APPROVED** [review]

**Comment by vshankar** [comment]
> > > LGTM.
> > > But do we have any idea how this can happen? This seems to be a different kind of problem than the other one triggered by mysql.
> > 
> > 
> > This instance did not have postgresql or any kind of database - the corruption happened to a random pdf or xml file :/
> 
> Yeah, but it was like 10% of files in the system, right? That seems like a bad logic path applying incorrect updates, not a memory corruption...

Right. I wonder why haven't we seen such corruption in our integration runs when the corruption is so widespread. What I/O pattern are we not testing in fs suite.

**Pull request merged**